### ISSUE TITLE
match_same_arms false positive with pattern guard

### ISSUE THREAD

**Issue by digama0** [rust-lang/rust-clippy#8637]
### Summary

The `match_same_arms` lint will suggest moving a match arm past a pattern guard, which is not semantics preserving and can cause an error if it produces an unreachable case.

### Lint Name

match_same_arms

### Reproducer

I tried this code:

```rust
#![warn(clippy::match_same_arms)]

fn main() {
  match None::<u8> {
    None => (),
    Some(n) if n < 2 => {}
    Some(_) => (),
  }
}
```

I saw this happen:

```
warning: this match arm has an identical body to another arm
 --> src/main.rs:7:5
  |
7 |     Some(_) => (),
  |     -------^^^^^^
  |     |
  |     help: try merging the arm patterns: `Some(_) | None`
  |
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![warn(clippy::match_same_arms)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^
  = help: or try changing either arm body
note: other arm here
 --> src/main.rs:5:5
  |
5 |     None => (),
  |     ^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#match_same_arms

warning: `rust-test` (bin "rust-test") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.89s
```

### Version

```text
rustc 1.62.0-nightly (60e50fc1c 2022-04-04)
binary: rustc
commit-hash: 60e50fc1cfe0bb693a5f4f93eb83ef70854531e3
commit-date: 2022-04-04
host: x86_64-unknown-linux-gnu
release: 1.62.0-nightly
LLVM version: 14.0.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by Jarcho** [comment]
This is not false-positive. The following is semantically equivalent:
```rust
match None::<u8> {
    Some(n) if n < 2 => {}
    Some(_) | None => (),
  }
```

The message should be clearer about which arm to modify. (I say this as the person who wrote it. I still don't particularly like it)

**Comment by digama0** [comment]
Aha, so it is. It's a bit awkward, but it wouldn't be the first time clippy makes a weird but technically correct suggestion.
